demos/fee-calculator/src/App.tsx

```diff
@@ -0,0 +1,225 @@
+import { useState } from 'react';
+import './App.css';
+import Parameter from './Parameter';
+import Output from './Output';
+import githubLogoUrl from './githubLogo.svg';
+
+const defaults = {
+  ethPrice: 2200,
+  l1GasPrice: 40,
+  l2GasPrice: 0.01,
+  l2CompressionRatio: 0.7,
+  calldataCostReduction4844: 1,
+  bundleSize: 10,
+  bundlerProfitMargin: 0.05,
+};
+
+const constants = {
+  transferGas: 21_000,
+  transferEffectiveBytes: 120,
+  gasPerByte: 16,
+  l2FixedOverhead: 188, // Fixed overhead of L1 gas (before compression)
+};
+
+const App = () => {
+  const [ethPrice, setEthPrice] = useState(defaults.ethPrice);
+  const [l1GasPrice, setL1GasPrice] = useState(defaults.l1GasPrice);
+  const [l2GasPrice, setL2GasPrice] = useState(defaults.l2GasPrice);
+
+  const [l2CompressionRatio, setL2CompressionRatio] = useState(
+    defaults.l2CompressionRatio,
+  );
+
+  const [calldataCostReduction4844, setCalldataCostReduction4844] = useState(
+    defaults.calldataCostReduction4844,
+  );
+
+  const [bundleSize, setBundleSize] = useState(defaults.bundleSize);
+
+  const [bundlerProfitMargin, setBundlerProfitMargin] = useState(
+    defaults.bundlerProfitMargin,
+  );
+
+  const l1TransferFee = ethPrice * l1GasPrice * 1e-9 * constants.transferGas;
+
+  const l2TransferFee = (() => {
+    const l1Gas =
+      (l2CompressionRatio / calldataCostReduction4844) *
+      (constants.transferEffectiveBytes * constants.gasPerByte +
+        constants.l2FixedOverhead);
+
+    const l2Gas = constants.transferGas;
+
+    return ethPrice * 1e-9 * (l1GasPrice * l1Gas + l2GasPrice * l2Gas);
+  })();
+
+  const l2TransferFee4337 = (() => {
```

Comment:
Trying to get my head around this expression as well. How did you arrive at the fixed values? (bundleOverheadEffectiveBytes - `155`, `455`, bundleOverheadGas - `25937`, `115195`)

Reply:
Yeah sorry about that. I need to clean this up a bit.

Anyway, 455 is the number of (effective) bytes in a bundle with one transfer. Therefore, including a transfer adds `455 - 155 = 300` bytes to the bundle. The bytes needed for a bundle even when it's empty need to be shared across all the user ops, which is why `effectiveBytesCharged` is `bundleOverheadEffectiveBytes / bundleSize + addedEffectiveBytes`.

`bundleOverheadGas` and `addedGas` are the same idea but for the gas native to the L2 (l2Gas). In the same way, an empty bundle needs a certain amount of gas, and then userOps expand that gas, so we charge fully for the added gas, and share the overhead needed for the bundle.

The constants involved for these are based on measurements taken using the in-page demo.

See previous answer for `l1Gas` calculation.

Reply:
Thanks for the clarification, all makes sense